### ISSUE TITLE
RSA keystore requires both public/private key to encrypt

### ISSUE THREAD

**Issue by minhster99** [spring-cloud/spring-cloud-config#1427]
Bug report

This bug is related to https://github.com/spring-cloud/spring-cloud-config/issues/539

According to the docs
"The encryption is done with the public key, and a private key is needed for decryption. Thus in principle you can configure only the public key in the server if you only want to do encryption (and are prepared to decrypt the values yourself locally with the private key)."

This implies the private key isn't required in the server keystore when used purely for encryption. I have created a keystore with only the public key in it and the server has picked it up. However when attempting to encrypt using the endpoint, I get the following exception

```
Servlet.service() for servlet [dispatcherServlet] in context with path [] threw exception [Request processing failed; nested exception is java.lang.IllegalStateException: Cannot load keys from store: class path resource [keystore-public.jks]] with root cause
java.lang.NullPointerException: null
	at org.springframework.security.rsa.crypto.KeyStoreKeyFactory.getKeyPair(KeyStoreKeyFactory.java:70)
	at org.springframework.cloud.config.server.encryption.KeyStoreTextEncryptorLocator.locate(KeyStoreTextEncryptorLocator.java:87)
	at org.springframework.cloud.config.server.encryption.EncryptionController.checkEncryptorInstalled(EncryptionController.java:166)
	at org.springframework.cloud.config.server.encryption.EncryptionController.encrypt(EncryptionController.java:126)
```

If my server keystore has both the private and public key in it, the problem goes away. We prefer not to do this from a security point of view as we want our config clients to perform the decryption and look after their own private keys.

**Comment by TYsewyn** [comment]
Hi @minhster99! Did you disable decryption in your `bootstrap.properties|yml` file of your config server by setting `spring.cloud.config.server.encrypt.enabled=false`?

**Comment by minhster99** [comment]
Hi @TYsewyn , yes I have that disabled but I wouldn't expect that to matter since I'm hitting the encrypt URL. I believe that only applies when you retrieve properties?

**Comment by TYsewyn** [comment]
> We prefer not to do this from a security point of view as we want our config clients to perform the decryption and look after their own private keys.

As per the docs:
> Sometimes you want the clients to decrypt the configuration locally, instead of doing it in the server. In that case, if you provide the `encrypt.*` configuration to locate a key, you can still have `/encrypt` and `/decrypt` endpoints, but you need to explicitly switch off the decryption of outgoing properties by placing `spring.cloud.config.server.encrypt.enabled=false` in `bootstrap.[yml|properties]`.

I assumed that setting this property would also configure the server to only use the public key, but apparently that's not the case.

**Comment by minhster99** [comment]
Ok I understand what you were getting at now. I haven't gotten that far yet :) 

Stuck at getting just the encryption to work on the server side without needing the private key.

**Comment by hviranicitco** [comment]
I see the same issue when I only have public key in the keystore. I also have spring.cloud.config.server.encrypt.enabled=false in bootstrap.

**Comment by TYsewyn** [comment]
@minhster99 @hviranicitco Can you both validate if this is working with the build snapshots of Greenwich and Hoxton?

**Comment by spencergibb** [comment]
Closed via #1518
